cmd/vic-init/toolbox.go

```diff
@@ -0,0 +1,178 @@
+	chsh.Wait()
+
+	// confirm the change
+	file, err := os.Open("/etc/passwd")
+	if err != nil {
+		err := fmt.Errorf("Failed to open file to confirm change: %s", err)
+		log.Error(err)
+		return err
+	}
+
+	reader := bufio.NewReader(file)
+	line, err := reader.ReadString('\n')
+	if err != nil {
+		err := fmt.Errorf("Failed to read line from file to confirm change: %s", err)
+		log.Error(err)
+		return err
+	}
+
+	// assert that first line is root
+	if !strings.HasPrefix(line, "root") {
+		err := fmt.Errorf("Expected line to start with root: %s", line)
+		log.Error(err)
+		return err
+	}
+
+	// assert that first line is root
+	if !strings.HasSuffix(line, "/bin/bash\n") {
+		err := fmt.Errorf("Expected line to end with /bin/bash: %s", line)
+		log.Error(err)
+		return err
+	}
+
+	return nil
+}
+
+// passwd sets the password for the root user to that provided as an argument
+func passwd(pass string) error {
+	defer trace.End(trace.Begin(""))
+
+	err := enableShell()
+	if err != nil {
+		err := fmt.Errorf("Failed to enable shell: %s", err)
+		log.Error(err)
+		// continue anyway as people may be able to get something useful
+	}
+
+	setPasswd := exec.Command("/sbin/chpasswd")
+	stdin, err := setPasswd.StdinPipe()
+	if err != nil {
+		err := fmt.Errorf("Failed to create stdin pipe for chpasswd: %s", err)
+		log.Error(err)
+		return err
+	}
+
+	err = setPasswd.Start()
+	if err != nil {
+		err := fmt.Errorf("Failed to launch chpasswd: %s", err)
+		log.Error(err)
+		return err
+	}
+
+	_, err = stdin.Write([]byte("root:" + pass))
+
+	// so that we're actively waiting when the process exits, or we'll race (and lose) to child reaper
+	var waitErr error
+	go func() {
+		waitErr = setPasswd.Wait()
+	}()
+
+	err = stdin.Close()
+	if err != nil {
+		err := fmt.Errorf("Failed to close input to chpasswd: %s", err)
+		log.Error(err)
+
+		// fire and forget as we're already on error path
+		setPasswd.Process.Kill()
+
+		return err
+	}
+
+	return nil
+}
+
+// enableSSH receives a key as an argument
+func enableSSH(key string) error {
+	defer trace.End(trace.Begin(""))
+
+	err := enableShell()
+	if err != nil {
+		err := fmt.Errorf("Failed to enable shell: %s", err)
+		log.Error(err)
+		// continue anyway as people may be able to get something useful
+	}
+
+	// basic sanity check for args - we don't bother validating it's a key
+	if len(key) != 0 {
+		err = os.MkdirAll("/root/.ssh", 0700)
+		if err != nil {
+			err := fmt.Errorf("unable to create path for keys: %s", err)
+			log.Error(err)
+			return err
+		}
+
+		err = ioutil.WriteFile("/root/.ssh/authorized_keys", []byte(key), 0600)
+		if err != nil {
+			err := fmt.Errorf("unable to create authorized_keys: %s", err)
+			log.Error(err)
+			return err
+		}
+	}
+
+	// because init is explicitly reaping child processes we cannot use simple
+	// exec commands to gather status
+	startSSH := exec.Command("/usr/bin/systemctl", "start", "sshd")
+	out, err := startSSH.CombinedOutput()
```

Comment:
Can't - there's no guarantee that Wait gets the exit value for the process given the child reaper. We just have to hope (or attempt to confirm as with the prior operations)